osmhelper/osm_connector.py

```diff
@@ -0,0 +1,325 @@
+            caching if possible.
+
+        :return stops: A dictionary of Stops constituting the obtained data.
+
+        """
+
+        # Preferably return cached data about stops
+        if refresh is False:
+            # Check if stops data is already built in this object
+            if not self.stops:
+                # If not, try to get stops data from file cache
+                self.stops = Cache.read_data('stops')
+            # Return cached data if found
+            if bool(self.stops):
+                return self.stops
+
+        # No cached data was found or refresh was forced
+        print "Query and build fresh data for stops"
+
+        # Obtain raw data about routes from OpenStreetMap
+        result = self._query_stops()
+
+        # Build stops from ways (polygons)
+        for stop in result.ways:
+            if Stop.is_valid_stop_candidate(stop):
+                self.stops["way/" + str(stop.id)
+                           ] = self._build_stop(stop, "way")
+
+        # Build stops from nodes
+        for stop in result.nodes:
+            if Stop.is_valid_stop_candidate(stop):
+                self.stops["node/" + str(stop.id)
+                           ] = self._build_stop(stop, "node")
+
+        # Cache and return whole data set
+        Cache.write_data('stops', self.stops)
+        return self.stops
+
+    def _build_route_master(self, route_master, members):
+        """Helper function to build a RouteMaster object
+
+        Returns a initiated RouteMaster object from raw data
+
+        """
+        if 'ref' in route_master.tags:
+            ref = route_master.tags['ref']
+        else:
+            sys.stderr.write(
+                "RouteMaster without 'ref': " + route_master.id + "\n")
+            print "http://osm.org/relation/" + + route_master.id
+            return
+        name = route_master.tags['name']
+        rm = RouteMaster(route_master.id, ref, name, members)
+        print rm
+        return rm
+
+    def _build_route_variant(self, route_variant, query_result_set):
+        """Helper function to build a RouteVariant object
+
+        Returns a initiated RouteVariant object from raw data
+
+        """
+        if 'ref' in route_variant.tags:
+            ref = route_variant.tags['ref']
+        else:
+            sys.stderr.write(
+                "RouteVariant without 'ref': " + route_variant.id + "\n")
+            print "http://osm.org/relation/" + + route_variant.id
+            return
+
+        stops = []
+
+        # Add ids for stops of this route variant
+        for stop_candidate in route_variant.members:
+            if stop_candidate.role == "platform":
+
+                if isinstance(stop_candidate, overpy.RelationNode):
+                    otype = "node"
+
+                elif isinstance(stop_candidate, overpy.RelationWay):
+                    otype = "way"
+
+                else:
+                    otype = "unknown"
```

Comment:
We use it right below and it's important!

> stops.append(otype + "/" + str(stop_candidate.ref))

It's needed because osm ids are not unique over nodes and ways. So, the id's we use are like `node/329857` and `way/329857`